src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+
+/**
+ ======= JSON Serialization and Streams =======
+ */
+namespace {
+static void to_json(const Bson& bson, std::stringstream& ss)
+{
+    std::visit(overloaded {
+        [&ss](auto) {
+            ss << "null";
+        },
+        [&ss](bson_null_t) {
+            ss << "null";
+        },
+        [&ss](int32_t bson) {
+            ss << "{" << "\"$numberInt\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](double bson) {
+            ss << "{" << "\"$numberDouble\"" << ":" << '"';
+            if (isnan(bson)) {
+                ss << "NaN";
+            } else if (bson == std::numeric_limits<double>::infinity()) {
+                ss << "Infinity";
+            } else if (bson == (-1 * std::numeric_limits<double>::infinity())) {
+                ss << "-Infinity";
+            } else {
+                ss << bson;
+            }
+            ss << '"' << "}";
+        },
+        [&ss](int64_t bson) {
+            ss << "{" << "\"$numberLong\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](Decimal128 bson) {
+            ss << "{" << "\"$numberDecimal\"" << ":" << '"';
+            if (bson.is_nan()) {
+                ss << "NaN";
+            } else if (bson == Decimal128("Infinity")) {
+                ss << "Infinity";
+            } else if (bson == Decimal128("-Infinity")) {
+                ss << "-Infinity";
+            } else {
+                ss << bson;
+            }
+            ss << '"' << "}";
+        },
+        [&ss](ObjectId bson) {
+            ss << "{" << "\"$oid\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&ss](BsonDocument bson) {
+            ss << "{";
+            for (auto const& pair : bson)
+            {
+                ss << '"' << pair.first << "\":";
+                to_json(pair.second, ss);
+                ss << ",";
+            }
+            if (bson.size())
+                ss.seekp(-1, std::ios_base::end);
+            ss << "}";
+        },
+        [&ss](BsonArray bson) {
+            ss << "[";
+            for (auto const& b : bson)
+            {
+                to_json(b, ss);
+                ss << ",";
+            }
+            if (bson.size())
+                ss.seekp(-1, std::ios_base::end);
+            ss << "]";
+        },
+        [&ss](BinaryData bson) {
+            ss << "{\"$binary\":{\"base64\":\"" << (bson.is_null() ? "" : bson.data()) << "\",\"subType\":\"00\"}}";
+        },
+        [&ss](RegularExpression bson) {
+            std::stringstream options;
+            for (const auto &option : bson.options()) options << option;
+            ss << "{\"$regularExpression\":{\"pattern\":\"" << bson.pattern()
+                << "\",\"options\":\"" << options.str() << "\"}}";
+        },
+        [&ss](Timestamp bson) {
+            ss << "{\"$timestamp\":{\"t\":" << bson.get_seconds() << ",\"i\":" << 1 << "}}";
```

Comment:
What is the "i" parameter?
Are we going to truncate the extra information we have stored in Timestamp::get_nanoseconds()?

Reply:
"i" is operations within a given second, which defaults to one.

We do not need the extra information– this field is meant to be internal to the bson implementation: https://docs.mongodb.com/manual/reference/bson-types/#timestamps